codegen: take gc roots (and alloca alignment) more seriously

vtjnash opened the pull request:
Due to limitations in the LLVM implementation, we are forced to emit fairly bad code here. But we need to make sure it is still correct with respect to GC rooting.

Fixes #54720

vtjnash committed:
codegen: take gc roots (and alloca alignment) more seriously

Due to limitations in the LLVM implementation, we are forced to emit
fairly bad code here. But we need to make sure it is still correct with
respect to GC rooting.

The PR 50833c84d454ef989797e035294ba27b3cca79b7 also changed the meaning
of haspadding without changing all of the existing users to use the new
equivalent flag (haspadding || !isbitsegal), incurring additional
breakage here as well and needing more tests for that.

Fixes #54720

Keno commented:
Can we get a test? At the very least a codegen test if the runtime behavior is non-deterministic.

gbaraldi commented:
Also maybe split this into two commits. One that does the seemingly NFC changes to use the `Align` type and the other with  the functional changes.

gbaraldi commented on src/cgutils.cpp:
```diff
@@ -2120,12 +2119,15 @@ static jl_cgval_t typed_store(jl_codectx_t &ctx,
+    bool tracked_pointers = isboxed || CountTrackedPointers(elty).count > 0;
     if (!isboxed && Order != AtomicOrdering::NotAtomic && !elty->isIntOrPtrTy()) {
+        intcast_eltyp = elty;
+        elty = Type::getIntNTy(ctx.builder.getContext(), 8 * nb);
```
Does this need a check so that N isn't too large? Or does the code that falls back to a lock avoid that?

vtjnash replied:
there is no limit on how large N can be per the standard (but yes, we limit to MAX_ATOMIC_SIZE because it generates better code)

vtjnash commented:
Added some tests to exercise this case, which was good because it found that https://github.com/JuliaLang/julia/pull/54121 had also introduced bugs in there recently when the spelling of the haspadding flag was changed to haspadding || !isbitsegal :/

vtjnash committed:
codegen: NFC refactoring to use Align type

vtjnash committed:
codegen: update type of x after type-assert

Later code likes to see that the type is consistent with the cgval and the unbox.

KristofferC commented:
I started backporting this but got a bit uncomfortable that I did things right. @vtjnash  maybe you could do it instead?

vtjnash commented:
Pushed to https://github.com/JuliaLang/julia/pull/55530. For future cases like this, note that Visual Studio Code was able to complete almost the entire merge automatically (where git itself failed quite a bit). Although not quite complete, since it also uncovered a bug, which I pushed to the backports branch and opened on master here: https://github.com/JuliaLang/julia/pull/55530
